ksql-engine/src/main/java/io/confluent/ksql/util/ExecutorWithRetries.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.util;
+
+import org.apache.kafka.common.errors.RetriableException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import java.util.function.Supplier;
+
+public final class ExecutorWithRetries {
+
+  private static final int NUM_RETRIES = 5;
+  private static final int RETRY_BACKOFF_MS = 500;
+  private static final Logger log = LoggerFactory.getLogger(ExecutorWithRetries.class);
+
+  private ExecutorWithRetries() {
+  }
+
+  public enum RetryBehaviour {
+    ALWAYS,
+    ON_RETRYABLE
+  }
+
+  @FunctionalInterface
+  public interface Function {
+    void call() throws Exception;
+  }
+
+  public static void execute(final Function function,
+                             final RetryBehaviour retryBehaviour) throws Exception {
+    execute(() -> {
+      function.call();
+      return null;
+    }, retryBehaviour);
+  }
+
+  public static <T> T execute(final Callable<T> supplier,
+                              final RetryBehaviour retryBehaviour) throws Exception {
+    return executeSync(() -> {
+      final CompletableFuture<T> f = new CompletableFuture<>();
+      try {
+        final T result = supplier.call();
+        f.complete(result);
+      } catch (final Exception e) {
+        f.completeExceptionally(e);
+      }
+      return f;
+    }, retryBehaviour);
+  }
+
+  public static <T> T executeSync(final Supplier<? extends Future<T>> supplier,
```

Comment:
I think the functions here can be simplified. It looks like we want two functions - one that takes a Supplier, and one that takes a Function, that call their argument in a retry loop and returns a value in the case of the Supplier. The fact that the admin client returns futures shouldn't leak into this class.

Reply:
Regarding the Future leaking in, are you suggesting changing the calling code from, say,

```
ExecutorWithRetries.execute(
   () -> kafkaTopicClient.deleteTopics(
   Collections.singletonList(dataSource.getKsqlTopic().getKafkaTopicName())),
   ExecutorWithRetries.RetryBehaviour.ALWAYS)
```

To:

```
ExecutorWithRetries.execute(
   () -> kafkaTopicClient.deleteTopics(
   Collections.singletonList(dataSource.getKsqlTopic().getKafkaTopicName())).get(),
   ExecutorWithRetries.RetryBehaviour.ALWAYS)
```

(Note the extra `get` call)

Meaning we can lose the `Future` part of the Executer?  That could work assuming the `deleteTopics` call itself never throws a retry-able exception, which is likely the case.

Reply:
deleteTopics doesn't return a future. All the calls that return a future call executeSync, which does ```supplier.get().get()```. If either the first or second get call throws a retryable exception there will be a retry.

Reply:
Sorry, my bad, wrong call.  But yes, that was what I was meaning. But you're right, the `Supplier<? extends Future<T>>` could just be a `Supplier<T>` with the `Future.get()` call made in the lambda passed to the executor.